File: backend/internal/port/user_repository.go
package port

import (
	"context"

	"github.com/sqwdvrt/siteParserForFreelans/backend/internal/domain"
)

// UserRepository — репозиторий пользователей.
type UserRepository interface {
	Save(ctx context.Context, telegramID int64) (userID int64, created bool, err error)
	GetByID(ctx context.Context, userID int64) (*domain.User, error)
	GetByTelegramID(ctx context.Context, telegramID int64) (*domain.User, error)
	GetPreferencesScoped(ctx context.Context, userID int64) (*domain.UserPreferences, error)
	UpdateProfile(ctx context.Context, userID int64, profileText string) error
	// UpdateProfileScoped выполняет обновление профиля в транзакции с установкой app.current_user_id для RLS.
	UpdateProfileScoped(ctx context.Context, userID int64, profileText string) error
	// UpdateNotifyHourScoped обновляет notify_hour в транзакции с установкой app.current_user_id для RLS.
	UpdateNotifyHourScoped(ctx context.Context, userID int64, hour int) error
	// UpsertPreferencesScoped обновляет user_preferences в транзакции с установкой app.current_user_id для RLS.
	UpsertPreferencesScoped(ctx context.Context, userID int64, prefs domain.UserPreferences) error
	// GetProUsersWithNotifyHour возвращает IDs pro-пользователей с заданным notify_hour (Europe/Moscow).
	GetProUsersWithNotifyHour(ctx context.Context, hour int) ([]int64, error)
}

File: ai-service/src/ai_service/port/user_repository.py
"""Port: UserRepository — абстракция доступа к users."""

from __future__ import annotations

from abc import ABC, abstractmethod

from ai_service.domain.user import User


class UserRepository(ABC):
    """Репозиторий для работы с пользователями."""

    @abstractmethod
    def list_matchable_users(self) -> list[User]:
        """Вернуть пользователей с embedding и preferences для pre-filter перед ANN."""
        ...

    @abstractmethod
    def save(self, telegram_id: int) -> int:
        """Создать пользователя. Возвращает user_id."""
        ...

    @abstractmethod
    def get_by_id(self, user_id: int) -> User | None:
        """Получить пользователя по id."""
        ...

    @abstractmethod
    def get_by_telegram_id(self, telegram_id: int) -> User | None:
        """Получить пользователя по telegram_id."""
        ...

    @abstractmethod
    def get_embedding(self, user_id: int) -> list[float] | None:
        """Получить embedding пользователя по id."""
        ...

    @abstractmethod
    def update_profile(self, user_id: int, profile_text: str) -> None:
        """Обновить profile_text пользователя."""
        ...

    @abstractmethod
    def save_embedding(self, user_id: int, embedding: list[float]) -> None:
        """Сохранить embedding пользователя."""
        ...
